### ISSUE TITLE
A Unified Styling Language  SEEK blog  Medium

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2739]
A Unified Styling Language – SEEK blog – Medium<br>
<br>
http://bit.ly/2rJST1L<br>
<br>
Mark Dalgleish<br>
<br>
<div><div>
<section name="70b3"><div><hr></div>
<div>
<div>
<div>
<div><a href="https://medium.com/@markdalgleish?source=post_header_lockup" dir="auto"><img src="https://cdn-images-1.medium.com/fit/c/100/100/1*psbV0S7m4-Al-OmOY_x30g.jpeg" alt="Go to the profile of Mark Dalgleish"></a></div>
<div>
<div>
<a href="https://medium.com/@markdalgleish" dir="auto">Mark Dalgleish</a><span></span>
</div>
<div>
<time datetime="2017-05-22T22:48:10.339Z">May 22, 2017</time><span></span><span title="22 min read"></span>
</div>
</div>
</div>
<p name="10ba" id="10ba">In the past few years we’ve seen the rise of <a href="https://github.com/MicheleBertoli/css-in-js" rel="nofollow noopener" target="_blank">CSS-in-JS</a>, emerging primarily from within the <a href="https://facebook.github.io/react" rel="nofollow noopener" target="_blank">React</a> community. This, of course, hasn’t been without its controversies. Many people, particularly those already intimately familiar with CSS, have looked on in disbelief.</p>
<blockquote name="115d" id="115d">“Why would anyone want to write CSS in JS?</blockquote>
<blockquote name="dbdb" id="dbdb">Surely this is a terrible idea!</blockquote>
<blockquote name="71b8" id="71b8">If only they’d learn CSS!”</blockquote>
<p name="42ce" id="42ce">If this was your reaction, then read on. We’re going to take a look at why writing your styles in JavaScript isn’t such a terrible idea after all, and why I think you should be keeping an eye on this rapidly evolving space.</p>
</div>
<div><figure name="8f5a" id="8f5a"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*Ipu5Grtzr21suPiTfvGXaw.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*Ipu5Grtzr21suPiTfvGXaw.png">
</div>
</div>
<figcaption>(<a href="https://github.com/airbnb/react-sketchapp" rel="nofollow noopener" target="_blank">Source</a>)</figcaption></figure></div>
<div>
<h3 name="482f" id="482f">Misunderstood communities</h3>
<p name="3f0f" id="3f0f">The React community is often misunderstood by the CSS community, and vice versa. This is particularly interesting for me, as I’m caught somewhere between these two worlds.</p>
<p name="e340" id="e340">I started learning HTML in the late nineties, and I’ve been working with CSS professionally since the dark ages of table-based layouts. Inspired by <a href="http://www.csszengarden.com" rel="nofollow noopener" target="_blank">CSS Zen Garden</a>, I was on the front lines of migrating existing codebases towards <a href="https://en.wikipedia.org/wiki/Semantic_HTML" rel="nofollow noopener" target="_blank">semantic markup</a> and cascading style sheets. It wasn’t long after this that I became obsessed with separating our concerns, using <a href="https://www.w3.org/wiki/The_principles_of_unobtrusive_JavaScript" rel="nofollow noopener" target="_blank">unobtrusive JavaScript</a> to decorate our server-rendered markup with client-side interactions. There was a small but vibrant community surrounding these practices, and we became the first generation of front-end developers, trying to give the browser platform the respect it deserved.</p>
<p name="f0a6" id="f0a6">With a web-focused background like this, you might imagine that I’d be vehemently opposed to React’s <a href="https://facebook.github.io/react/docs/jsx-in-depth.html" rel="nofollow noopener" target="_blank"><em>HTML-in-JS</em></a> model, which seemingly goes against the principles we held so dear—but in fact, it’s quite the opposite. In my experience, React’s component model, coupled with its ability to render server-side, finally gives us a way to build complex single-page apps at scale, in a way that still allows us to ship fast, accessible, progressively enhanced products to our users. We’ve even leveraged this ability here at <a href="https://www.seek.com.au" rel="nofollow noopener" target="_blank">SEEK</a>, our flagship product being a single-page React app where the core search flow still works when JavaScript is disabled—gracefully degrading to a traditional web site by running the same JavaScript code on the server.</p>
<p name="01c6" id="01c6">So, consider this an olive branch from one community to another. Together, let’s try and understand what this movement is all about. It might not be <em>perfect,</em> it might not be something you plan to use in your products, it might not even be terribly convincing for you—but it’s at least worth trying to wrap your head around.</p>
<h3 name="c9e8" id="c9e8">Why CSS-in-JS?</h3>
<p name="eacc" id="eacc">If you’re familiar with my recent work with React and <a href="https://github.com/css-modules/css-modules" rel="nofollow noopener" target="_blank">CSS Modules</a>, you may be surprised to see me defending CSS-in-JS.</p>
<figure name="c46a" id="c46a"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*RtAMWbxdwW2ujyrurU9plw.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*RtAMWbxdwW2ujyrurU9plw.png">
</div>
</div>
<figcaption>CSS Modules (<a href="https://github.com/css-modules/css-modules" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="b5ee" id="b5ee">After all, CSS Modules is typically chosen by developers who want locally-scoped styles <em>without </em>buying in to CSS-in-JS. In fact, I don’t even use CSS-in-JS in my own work.</p>
<p name="e417" id="e417">Despite this, I continue to maintain a keen interest in the CSS-in-JS community, keeping a close eye on the innovations that they continually come up with. Not only that, <em>I think the broader CSS community should be interested too.</em></p>
<p name="cf0c" id="cf0c">But why?</p>
<p name="1213" id="1213">To get a clearer understanding of why people are choosing to write their styles in JavaScript, we’ll focus on the practical benefits that emerge when taking this approach.</p>
<p name="c5f2" id="c5f2">I’ve broken this down into five major areas:</p>
<ol>
<li name="f512" id="f512">Scoped styles</li>
<li name="3a67" id="3a67">Critical CSS</li>
<li name="ac29" id="ac29">Smarter optimisations</li>
<li name="5696" id="5696">Package management</li>
<li name="bfa0" id="bfa0">Non-browser styling</li>
</ol>
<p name="a62f" id="a62f">Let’s break this down further and have a closer look at what CSS-in-JS brings to the table for each of these points.</p>
</div>
</div></section><section name="33e1"><div><hr></div>
<div><div>
<h3 name="e60e" id="e60e">1.</h3>
<h4 name="da72" id="da72"><strong>Scoped styles</strong></h4>
<p name="f46c" id="f46c">It’s no secret that architecting CSS effectively at scale is incredibly difficult. When joining an existing long-lived project, it wasn’t uncommon to find that the CSS was the hardest part of the system to figure out.</p>
<p name="5cd8" id="5cd8">To counter this, the CSS community has invested heavily in trying to address these issues, making our styles more maintainable with methodologies like <a href="https://github.com/stubbornella/oocss/wiki" rel="noopener nofollow" target="_blank">OOCSS</a> by <a href="https://twitter.com/stubbornella" rel="nofollow noopener" target="_blank">Nicole Sullivan</a> and <a href="https://smacss.com/" rel="noopener nofollow" target="_blank">SMACSS</a> by <a href="https://twitter.com/snookca" rel="noopener nofollow" target="_blank">Jonathan Snook</a>—but the clear winner right now in terms of popularity seems to be <a href="http://getbem.com" rel="noopener nofollow" target="_blank">BEM</a>, or Block Element Modifier, from <a href="https://github.com/yandex" rel="noopener nofollow" target="_blank">Yandex</a>.</p>
<p name="8f90" id="8f90">Ultimately, BEM (when applied purely to CSS) is just a naming convention, opting to limit styles to classes following a <code>.Block__element--modifier</code> pattern. In any given BEM-styled codebase, developers have to remember to follow BEM’s rules at all times. When strictly followed, BEM works really well, but why is something as fundamental as scoping left up to pure <em>convention?</em></p>
<p name="acd8" id="acd8">Whether they explicitly say so or not, the majority of CSS-in-JS libraries are following the BEM mindset of trying to target styles to individual UI elements, but implementing it in a radically different way.</p>
<p name="0a55" id="0a55">What does this look like in practice? When using <a href="https://github.com/threepointone/glamor" rel="nofollow noopener" target="_blank">glamor</a> by <a href="https://twitter.com/threepointone" rel="nofollow noopener" target="_blank">Sunil Pai</a>, it looks something like this:</p>
<pre name="cc1e" id="cc1e">import { css } from 'glamor'</pre>
<pre name="0208" id="0208">const title = <strong>css({<br>  fontSize: '1.8em',<br>  fontFamily: 'Comic Sans MS',<br>  color: 'blue'<br>})</strong></pre>
<pre name="5487" id="5487">console.log(<strong>title</strong>)</pre>
<pre name="956c" id="956c">// → <strong>'css-1pyvz'</strong></pre>
<p name="1480" id="1480">What you’ll notice here is that <em>the CSS class is nowhere to be found in our code.</em> It’s no longer a hard-coded reference to a class defined elsewhere in the system. Instead, it is generated automatically for us by the library. We don’t have to worry about our selectors clashing in the global scope, which means we no longer have to manually prefix them.</p>
<p name="9237" id="9237">The scoping of this selector matches the scoping rules of the surrounding code. If you want to make this rule available to the rest of your application, you’ll need to turn it into a JavaScript module and import it wherever it’s used. In terms of keeping our codebases maintainable over time, this is incredibly powerful, <em>making sure that the source of any given style can be easily traced like any other code.</em></p>
<p name="6702" id="6702"><strong><em>By moving from mere convention towards enforcing locally-scoped styles by default, we’ve now improved the baseline quality of our styles. BEM is baked in, not opt-in.</em></strong></p>
<p name="ed9d" id="ed9d">—</p>
<p name="fee4" id="fee4">Before I continue, there’s a critically important point to clarify.</p>
<p name="6ce3" id="6ce3"><strong><em>This is generating real CSS, not inline styles.</em></strong></p>
<p name="deb1" id="deb1">Most of the earliest CSS-in-JS libraries attached styles directly to each element, but the critical flaw in this model is that ‘style’ attributes can’t do everything that CSS can do. Most new libraries instead focus on <em>dynamic style sheets,</em> inserting and removing rules at runtime from a global set of styles.</p>
<p name="6667" id="6667">As an example, let’s have a look at <a href="https://github.com/cssinjs/jss" rel="nofollow noopener" target="_blank">JSS</a> by <a href="https://twitter.com/oleg008" rel="nofollow noopener" target="_blank">Oleg Slobodskoi</a>, one of the earliest CSS-in-JS libraries to generate <em>real CSS.</em></p>
<figure name="4cbe" id="4cbe"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*ltBvwbyvBt8OMdGZQOdMDA.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*ltBvwbyvBt8OMdGZQOdMDA.png">
</div>
</div>
<figcaption>JSS (<a href="https://github.com/cssinjs/logo" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="58a9" id="58a9">When using JSS, you can use standard CSS features like hover styles and media queries, which map directly to the equivalent CSS rules.</p>
<pre name="39e8" id="39e8">const styles = {<br>  button: {<br>    padding: '10px',<br><strong>'&amp;:hover':</strong> {<br>      background: 'blue'<br>    }<br>  },<br><strong>'@media (min-width: 1024px)':</strong> {<br>    button: {<br>      padding: '20px'<br>    }<br>  }<br>}</pre>
<p name="bc90" id="bc90">Once you insert these styles into the document, the automatically generated classes are provided to you.</p>
<pre name="19bd" id="19bd">const { <strong>classes</strong> } = jss.createStyleSheet(<strong>styles</strong>).attach()</pre>
<p name="b746" id="b746">These generated classes can then be used instead of hard-coded class strings when generating markup in JavaScript. This pattern works regardless of whether you’re using a full-blown framework, or something as simple as <em>innerHTML.</em></p>
<pre name="62c9" id="62c9">document.body.innerHTML = `<br>  &lt;h1 class="${<strong>classes.heading</strong>}"&gt;Hello World!&lt;/h1&gt;<br>`</pre>
<p name="a540" id="a540">Managing the styles in this way is of little benefit by itself—it’s usually paired with some kind of component library. As a result, it’s typical to find bindings available for the most popular libraries. For example, JSS can easily bind to React components with the help of <a href="https://github.com/cssinjs/react-jss" rel="nofollow noopener" target="_blank">react-jss</a>, injecting a small set of styles into each component while managing the global lifecycle for you.</p>
<pre name="0957" id="0957">import injectSheet from 'react-jss'</pre>
<pre name="af93" id="af93">const Button = ({ <strong>classes</strong>, children }) =&gt; (<br>  &lt;button className={<strong>classes.button</strong>}&gt;<br>    &lt;span className={<strong>classes.label</strong>}&gt;<br>      {children}<br>    &lt;/span&gt;<br>  &lt;/button&gt;<br>)</pre>
<pre name="ec62" id="ec62">export default <strong>injectSheet(styles)(Button)</strong></pre>
<p name="c418" id="c418">By focusing our styles around components, integrating them more closely at the code level, we’re effectively taking BEM to its logical conclusion. So much so that many in the CSS-in-JS community felt like the importance of extracting, naming and reusing components was being lost in all the style-binding boilerplate.</p>
<p name="0241" id="0241">An entirely new way of thinking about this problem emerged with the introduction of <a href="https://github.com/styled-components/styled-components" rel="nofollow noopener" target="_blank">styled-components</a> by <a href="https://twitter.com/glenmaddern" rel="nofollow noopener" target="_blank">Glen Maddern</a> and <a href="https://twitter.com/mxstbr" rel="nofollow noopener" target="_blank">Max Stoiber</a>.</p>
<figure name="c83a" id="c83a"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*l4nfMFKxfT4yNTWUK2Vsdg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*l4nfMFKxfT4yNTWUK2Vsdg.png">
</div>
</div>
<figcaption>Styled Components (<a href="https://github.com/styled-components/styled-components" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="00d5" id="00d5">Instead of creating styles, which we then have to manually bind to components, we’re forced to create components directly.</p>
<pre name="f063" id="f063">import styled from 'styled-components'<br><br>const <strong>Title</strong> = styled.<strong>h1</strong>`<br><strong>font-family: Comic Sans MS;<br>  color: blue;</strong><br>`</pre>
<p name="432f" id="432f">When applying these styles, we don’t attach a class to an existing element. We simply render the generated component.</p>
<pre name="5c39" id="5c39"><strong>&lt;Title&gt;</strong>Hello World!<strong>&lt;/Title&gt;</strong></pre>
<p name="eb87" id="eb87">While styled-components uses traditional CSS syntax via tagged template literals, others prefer working with data structures instead. A notable alternative is <a href="https://github.com/paypal/glamorous" rel="nofollow noopener" target="_blank">Glamorous</a> by <a href="https://twitter.com/kentcdodds" rel="nofollow noopener" target="_blank">Kent C. Dodds</a> from <a href="https://github.com/paypal" rel="nofollow noopener" target="_blank">PayPal</a>.</p>
<figure name="44b0" id="44b0"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*Ere9GQTIJeNac2ONfZlfdw.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*Ere9GQTIJeNac2ONfZlfdw.png">
</div>
</div>
<figcaption>Glamorous (<a href="https://github.com/paypal/glamorous" rel="nofollow noopener" target="_blank">Source</a>)</figcaption></figure><p name="40ed" id="40ed">Glamorous offers the same component-first API as styled-components, but opts for <em>objects</em> instead of <em>strings,</em> eliminating the need to include a CSS parser in the library—reducing the library’s size and performance footprint.</p>
<pre name="de02" id="de02">import glamorous from 'glamorous'<br><br>const Title = glamorous.h1(<strong>{<br>  fontFamily: 'Comic Sans MS',<br>  color: 'blue'<br>}</strong>)</pre>
<p name="5869" id="5869">Whichever syntax you use to describe your styles, they’re no longer just <em>scoped</em> to components—<em>they’re inseparable from them.</em> When using a library like React, components are the fundamental building blocks, and now our styles form a core part of that architecture. <em>If we describe everything in our app as components, why not our styles too?</em></p>
<p name="f455" id="f455">—</p>
<p name="0058" id="0058">For seasoned veterans of BEM, all of this may seem like a relatively shallow improvement given the significance of the change we’ve introduced to our system. In fact, CSS Modules lets you achieve this without leaving the comfort of the CSS tooling ecosystem. This is why so many projects stick with CSS Modules, finding that it sufficiently solves most of their problems with writing CSS at scale without sacrificing the familiarity of regular CSS.</p>
<p name="da0e" id="da0e">However, it’s when we start to build on top of these basic concepts that things start to get a lot more interesting.</p>
</div></div></section><section name="66c6"><div><hr></div>
<div><div>
<h3 name="9619" id="9619">2.</h3>
<h4 name="6de3" id="6de3">Critical CSS</h4>
<p name="7d45" id="7d45">It’s become a relatively recent best practice to inline critical styles in the head of your document, improving initial load times by providing only those styles required to render the current page. This is in stark contrast to how we usually loaded styles—forcing the browser to download every possible visual style for our application before a single pixel was rendered on the screen.</p>
<p name="c566" id="c566">While there are tools available for extracting and inlining critical CSS, like the appropriately named <a href="https://github.com/addyosmani/critical" rel="nofollow noopener" target="_blank">critical</a> by <a href="https://twitter.com/addyosmani" rel="nofollow noopener" target="_blank">Addy Osmani</a>, they don’t fundamentally change the fact that critical CSS is difficult to maintain and automate. It’s a tricky, purely optional performance optimisation, so most projects seem to forgo this step.</p>
<p name="2897" id="2897">CSS-in-JS is a totally different story.</p>
<p name="9a8c" id="9a8c">When working in a server rendered application, extracting your critical CSS is not merely an optimisation—CSS-in-JS on the server fundamentally <em>requires</em> critical CSS to even work in the first place.</p>
<p name="4474" id="4474">For example, when using <a href="https://github.com/Khan/aphrodite" rel="nofollow noopener" target="_blank">Aphrodite</a> from <a href="https://github.com/Khan" rel="nofollow noopener" target="_blank">Khan Academy</a>, it keeps track of which styles are used within a single render pass using its <code>css</code> function, which is called inline while applying classes to your elements.</p>
<pre name="39f1" id="39f1">import { StyleSheet, <strong>css</strong> } from 'aphrodite'</pre>
<pre name="e6b8" id="e6b8">const styles = StyleSheet.create({<br>  title: { ... }<br>})</pre>
<pre name="a688" id="a688">const Heading = ({ children }) =&gt; (<br>  &lt;h1 className={<strong>css(styles.heading)</strong>}&gt;{ children }&lt;/h1&gt;<br>)</pre>
<p name="c8d1" id="c8d1">Even though all of your styles are defined in JavaScript, you can easily extract all the styles for the current page into a static string of CSS that can be inserted into the head of the document when rendering server-side.</p>
<pre name="7f6a" id="7f6a">import { StyleSheetServer } from 'aphrodite';<br><br>const { html, <strong>css</strong> } = <strong>StyleSheetServer.renderStatic</strong>(() =&gt; {<br>  return ReactDOMServer.renderToString(&lt;App/&gt;);<br>});</pre>
<p name="44ef" id="44ef">You can now render your critical CSS block like this:</p>
<pre name="be04" id="be04">const criticalCSS = `<br>  &lt;style data-aphrodite&gt;<br><strong>${css.content}</strong><br>  &lt;/style&gt;<br>`;</pre>
<p name="bba5" id="bba5">If you’ve looked into React’s server rendering model, you may find this to be a very familiar pattern. In React, your components define their markup in JavaScript, but can be rendered to a regular HTML string on the server.</p>
<p name="682e" id="682e"><strong><em>If you build your app with progressive enhancement in mind, despite being written entirely in JavaScript, it might not require JavaScript on the client at all.</em></strong></p>
<p name="5d8c" id="5d8c">Either way, the client-side JavaScript bundle includes the code needed to boot up your single-page app, suddenly bringing it to life, rendering in the browser from that point forward.</p>
<p name="d9f3" id="d9f3">Since rendering your HTML and CSS on the server happen at the same time, libraries like Aphrodite often help us streamline the generation of both critical CSS <em>and</em> server-rendered HTML into a single call, as we saw in the previous example. This now allows us to render our React components to static HTML in a similar fashion.</p>
<pre name="34c5" id="34c5">const appHtml = `<br>  &lt;div id="root"&gt;<br><strong>${html}</strong><br>  &lt;/div&gt;<br>`</pre>
<p name="ff9a" id="ff9a">By using CSS-in-JS on the server, not only does our single-page app continue to work without JavaScript, <em>it might even render faster too.</em></p>
<p name="834b" id="834b"><strong><em>As with the scoping of our selectors, the best practice of rendering critical CSS is now baked in, not opt-in.</em></strong></p>
</div></div></section><section name="51f6"><div><hr></div>
<div><div>
<h3 name="ac7f" id="ac7f">3.</h3>
<h4 name="7346" id="7346">Smarter optimisations</h4>
<p name="bd3f" id="bd3f">We’ve recently seen the rise of new ways of structuring our CSS—like <a href="https://acss.io/" rel="nofollow noopener" target="_blank">Atomic CSS</a> from <a href="https://github.com/yahoo" rel="nofollow noopener" target="_blank">Yahoo</a> and <a href="http://tachyons.io/" rel="nofollow noopener" target="_blank">Tachyons</a> by <a href="https://twitter.com/mrmrs_" rel="nofollow noopener" target="_blank">Adam Morse</a>—that eschew “semantic classes” in favour of tiny, single-purpose classes. For example, when using Atomic CSS, you apply classes with a function-like syntax which can then be used to generate an appropriate style sheet.</p>
<pre name="1c4d" id="1c4d">&lt;div class="<strong>Bgc(#0280ae.5) C(#fff) P(20px)</strong>"&gt;<br>  Atomic CSS<br>&lt;/div&gt;</pre>
<p name="69f1" id="69f1">The goal is to keep your CSS bundle as lean as possible by maximising the re-usability of classes, effectively treating classes like inline styles. While it’s easy to appreciate the reduction in file size, the impacts to both your codebase and your fellow team members are anything but insignificant. These optimisations, by their very nature, involve changes to both your CSS <em>and</em> your markup, making them a more significant architectural effort.</p>
<p name="a7d3" id="a7d3">As we’ve touched on already, when using CSS-in-JS or CSS Modules, you no longer hard-code class strings in your markup, instead using dynamic references to JavaScript values that have been generated automatically by a library or build tool.</p>
<p name="e12d" id="e12d">Instead of this:</p>
<pre name="c960" id="c960">&lt;aside className="sidebar" /&gt;</pre>
<p name="188e" id="188e">We write this:</p>
<pre name="05c9" id="05c9">&lt;aside className={styles.sidebar} /&gt;</pre>
<p name="eb92" id="eb92">This may look like a fairly superficial change, but this is a monumental shift in how we manage the relationship between our markup and our styles. By giving our CSS tooling the ability to alter not just our styles, <em>but the final classes we apply to our elements,</em> we unlock an entirely new class of optimisations for our style sheets.</p>
<p name="e842" id="e842">If we look at the example above, <em>‘styles.sidebar’</em> evaluates to a string, but there’s nothing limiting it to a single class. For all we know, it could just as easily end up being a string of over a dozen classes.</p>
<pre name="4543" id="4543">&lt;aside className={styles.sidebar} /&gt;</pre>
<pre name="702d" id="702d">// Could easily resolve to this:</pre>
<pre name="e61f" id="e61f">&lt;aside className={<strong>'class1 class2 class3 class4'</strong>} /&gt;</pre>
<p name="d0fd" id="d0fd">If we can optimise our styles, generating multiple classes for each set of styles, we can do some really interesting things.</p>
<p name="1f2e" id="1f2e">My favourite example of this is <a href="https://github.com/rtsao/styletron" rel="nofollow noopener" target="_blank">Styletron</a> by <a href="https://twitter.com/rtsao" rel="nofollow noopener" target="_blank">Ryan Tsao</a>.</p>
<figure name="f9aa" id="f9aa"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*7xxb6FOmcmPCnQNrFy5pjg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*7xxb6FOmcmPCnQNrFy5pjg.png">
</div>
</div>
<figcaption>Styletron (<a href="https://github.com/rtsao/styletron" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="4511" id="4511">In the same way that CSS-in-JS and CSS Modules automate the process of adding BEM-style class prefixes, Styletron does the same thing to the Atomic CSS mindset.</p>
<p name="839d" id="839d">The core API is focused around a single task—defining individual CSS rules for each combination of property, value and media query, which then returns an automatically generated class.</p>
<pre name="14e6" id="14e6">import styletron from 'styletron';</pre>
<pre name="bbcd" id="bbcd">styletron.injectDeclaration(<strong>{<br>  prop: 'color',<br>  val: 'red',<br>  media: '(min-width: 800px)'<br>}</strong>);</pre>
<pre name="9800" id="9800">// → 'a'</pre>
<p name="6f62" id="6f62">Of course, Styletron provides higher level APIs, such as its <code>injectStyle</code> function which allows multiple rules to be defined at once.</p>
<pre name="9bbc" id="9bbc">import { injectStyle } from 'styletron-utils';</pre>
<pre name="bc6e" id="bc6e">injectStyle(styletron, {<br><strong>color: 'red',</strong><br>  display: 'inline-block'<br>});<br>// → '<strong>a</strong> d'</pre>
<pre name="1a59" id="1a59">injectStyle(styletron, {<br><strong>color: 'red',</strong><br>  fontSize: '1.6em'<br>});<br>// → '<strong>a</strong> e'</pre>
<p name="410b" id="410b">Take special note of the commonality between the two sets of class names generated above.</p>
<p name="b25a" id="b25a"><strong><em>By relinquishing low-level control over the classes themselves, only defining the desired set of styles, we allow the library to generate the optimal set of atomic classes on our behalf.</em></strong></p>
<figure name="e36d" id="e36d"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*pWXr1A6uhiOkYHqwfBMtWg.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*pWXr1A6uhiOkYHqwfBMtWg.png">
</div>
</div>
<figcaption>CSS-in-JS bundle size comparison using Airbnb’s styles (<a href="https://ryantsao.com/blog/virtual-css-with-styletron" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="85b0" id="85b0">Optimisations that are typically done by hand—finding the most efficient way to split up our styles into reusable classes—can now be completely automated. You might be starting to notice a trend here. <strong><em>Atomic CSS is baked in, not opt-in.</em></strong></p>
</div></div></section><section name="1091"><div><hr></div>
<div><div>
<h3 name="82c1" id="82c1">4.</h3>
<h4 name="d1ec" id="d1ec">Package management</h4>
<p name="f688" id="f688">Before digging into this point, it’s first worth stopping and asking yourself a seemingly simple question.</p>
<p name="103f" id="103f"><em>How do we share CSS with each other?</em></p>
<p name="46a7" id="46a7">We’ve migrated from manually downloading CSS files, towards front-end specific package managers like <a href="https://bower.io" rel="nofollow noopener" target="_blank">Bower</a>, and now via <a href="https://www.npmjs.com/" rel="nofollow noopener" target="_blank">npm</a> thanks to tools like <a href="http://browserify.org/" rel="nofollow noopener" target="_blank">Browserify</a> and <a href="https://webpack.js.org" rel="nofollow noopener" target="_blank">webpack</a>. Even though some tooling has automated the process of including CSS from external packages, the front-end community has mostly settled on manual inclusion of CSS dependencies.</p>
<p name="0a38" id="0a38">Either way, there’s one thing that CSS dependencies aren’t very good at—depending on <em>other</em> CSS dependencies.</p>
<p name="5be9" id="5be9">As many of you might remember, we’ve seen a similar effect with <em>JavaScript modules</em> between Bower and npm.</p>
<p name="88f3" id="88f3">Bower wasn’t coupled to any particular module format, whereas modules published to npm use the <a href="http://wiki.commonjs.org/wiki/Modules/1.1" rel="nofollow noopener" target="_blank">CommonJS module format</a>. This has had a massive impact on the number of packages published to each platform.</p>
<p name="9274" id="9274">Complex trees of small, <em>nested</em> dependencies felt right at home on npm, while Bower attracted large, <em>monolithic</em> dependencies, of which you might only have two or three—plus a few plugins, of course. Since your dependencies didn’t have a module system to rely on, each package couldn’t easily make use of its <em>own</em> dependencies, so the integration was always a manual step left up to the consumer.</p>
<p name="9c57" id="9c57">As a result, the number of packages on npm over time forms an exponential curve, while Bower only had a fairly linear increase of packages. While there are certainly a variety of reasons for this differentiation, it’s fair to say that a <em>lot</em> of it has to do with the way each platform allows (or doesn’t allow) packages to depend on each other at runtime.</p>
<figure name="b13a" id="b13a"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*LTrsIISPV5qK-qAQKaeINA.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*LTrsIISPV5qK-qAQKaeINA.png">
</div>
</div>
<figcaption>Yes, it’s this graph again. (<a href="https://www.quora.com/Why-use-Bower-when-there-is-npm/answer/Mattias-Petter-Johansson" rel="nofollow noopener" target="_blank">Source</a>)</figcaption></figure><p name="9e10" id="9e10">Unfortunately, this looks all too familiar to the CSS community, where we’ve also seen a relatively slow increase of monolithic packages compared to what we see with JavaScript packages on npm.</p>
<p name="cd77" id="cd77">What if we instead wanted to match npm’s exponential growth? What if we wanted to be able to depend on complex package hierarchies of varying sizes, with less focus on large, all-encompassing frameworks? To do this, we’d not only need a package manager that’s up to the task—we’d also need an appropriate module format too.</p>
<p name="c207" id="c207">Does this mean that we need a package manager specifically designed for CSS? For preprocessors like Sass and Less?</p>
<p name="9d4e" id="9d4e">What’s really interesting is that we’ve already gone through a similar realisation with HTML. If you ask the same questions about how we share <em>markup</em> with each other, you’ll quickly notice that we almost never share raw HTML directly—we share <em>HTML-in-JS.</em></p>
<p name="5500" id="5500">We do this via <a href="https://plugins.jquery.com/" rel="nofollow noopener" target="_blank">jQuery plugins</a>, <a href="http://ngmodules.org" rel="nofollow noopener" target="_blank">Angular directives</a> and <a href="https://react.parts/web" rel="nofollow noopener" target="_blank">React components</a>. We compose large components out of smaller components, each with their own HTML, each published independently to npm. HTML as a format might not be powerful enough to allow this, but by <em>embedding HTML within a fully-fledged programming language,</em> we’re easily able to work around this limitation.</p>
<p name="afe2" id="afe2">What if, like HTML, we shared our CSS—and the logic that generates it—via JavaScript? What if, instead of using <a href="http://sass-lang.com/documentation/file.SASS_REFERENCE.html#mixins" rel="nofollow noopener" target="_blank">mixins</a>, we used <em>functions returning objects and strings?</em> Instead of <a href="http://sass-lang.com/documentation/file.SASS_REFERENCE.html#extend" rel="nofollow noopener" target="_blank">extending classes</a>, what if we simply <em>merged objects</em> with <a href="https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign" rel="nofollow noopener" target="_blank"><em>Object.assign</em></a>, or the new <a href="https://github.com/tc39/proposal-object-rest-spread" rel="nofollow noopener" target="_blank">object spread operator</a>?</p>
<pre name="990d" id="990d">const styles = {<br><strong>  ...rules,<br>  ...moreRules,<br></strong>  fontFamily: 'Comic Sans MS',<strong><br></strong> color: 'blue'<br>}</pre>
<p name="8d08" id="8d08">Once we start writing our styles this way, we can now compose and share our styling code like any other code in our application, using the same patterns, the same tooling, the same infrastructure, <em>the same ecosystem.</em></p>
<p name="555b" id="555b">A great example of how this starts to pay off is in libraries like <a href="https://github.com/styled-components/polished" rel="nofollow noopener" target="_blank">Polished</a> by <a href="https://twitter.com/mxstbr" rel="nofollow noopener" target="_blank">Max Stoiber</a>, <a href="https://twitter.com/nikgraf" rel="nofollow noopener" target="_blank">Nik Graf</a> and <a href="https://twitter.com/b_hough" rel="nofollow noopener" target="_blank">Brian Hough</a>.</p>
<figure name="af44" id="af44"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*fczf3OWmmKBkFgtUqZnq2g.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*fczf3OWmmKBkFgtUqZnq2g.png">
</div>
</div>
<figcaption>Polished (<a href="https://github.com/styled-components/polished" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="35bd" id="35bd">Polished is essentially the <a href="https://lodash.com" rel="nofollow noopener" target="_blank">Lodash</a> of CSS-in-JS, providing a complete suite of mixins, colour functions, shorthands and more, making the process of authoring styles in JavaScript much more familiar for those coming from languages like <a href="http://sass-lang.com" rel="nofollow noopener" target="_blank">Sass</a>. The key difference now is that this code is much more composable, testable and shareable, able to use the full JavaScript package ecosystem.</p>
<p name="f185" id="f185">So, when it comes to CSS, how do we achieve the same level of open source activity seen across npm as a whole, composing large collections of styles from small, reusable, open-source packages? Strangely enough, we might finally achieve this by embedding our CSS within another language and completely embracing JavaScript modules.</p>
</div></div></section><section name="e384"><div><hr></div>
<div><div>
<h3 name="bcba" id="bcba">5.</h3>
<h4 name="7f92" id="7f92">Non-browser styling</h4>
<p name="9ecf" id="9ecf">So far, all of the points we’ve covered—while certainly a lot <em>easier</em> when writing CSS in JavaScript—they’re by no means things that are <em>impossible</em> with regular CSS. This is why I’ve left the most interesting, future-facing point until last. Something that, while not necessarily playing a huge role in today’s CSS-in-JS community, is quite possibly going to become a foundational layer in the future of <em>design.</em> Something that affects not just developers, but designers too, radically altering the way these two disciplines communicate with each other.</p>
<p name="6a03" id="6a03">First, to put this in context, we need to take a quick detour into React.</p>
<p name="f574" id="f574">—</p>
<p name="a1c5" id="a1c5">The React model is all about components rendering intermediate representations of the final output. When working in the browser, rather than directly mutating DOM elements, we’re building up complex trees of virtual DOM.</p>
<p name="0feb" id="0feb">What’s interesting, though, is that rendering to the DOM is not part of the core React library, instead being provided by <em>react-dom.</em></p>
<pre name="b186" id="b186">import { render } from <strong>'react-dom'</strong></pre>
<p name="0a58" id="0a58">Even though React was built for the DOM first, and still gets most of its usage in that environment, this model allows React to target wildly different environments simply by introducing new renderers.</p>
<p name="a305" id="a305">JSX isn’t just about virtual DOM—it’s about virtual <em>whatever.</em></p>
<p name="8732" id="8732">This is what allows <a href="https://facebook.github.io/react-native" rel="nofollow noopener" target="_blank">React Native</a> to work, writing truly native applications in JavaScript, by writing components that render virtual representations of their native counterparts. Instead of <em>div</em> and <em>span</em>, we have <em>View</em> and <em>Text.</em></p>
<p name="b656" id="b656">From a CSS perspective, the most interesting thing about React Native is that it comes with its own <a href="https://facebook.github.io/react-native/docs/stylesheet.html" rel="nofollow noopener" target="_blank">StyleSheet API</a>:</p>
<pre name="5e42" id="5e42">var styles = <strong>StyleSheet.create</strong>({<br>  container: {<br>    borderRadius: 4,<br>    borderWidth: 0.5,<br>    borderColor: '#d6d7da',<br>  },<br>  title: {<br>    fontSize: 19,<br>    fontWeight: 'bold',<br>  },<br>  activeTitle: {<br>    color: 'red',<br>  }<br>})</pre>
<p name="a8d8" id="a8d8">Here you can see a familiar set of styles, in this case covering colours, fonts and border styling.</p>
<p name="04fa" id="04fa">These rules are fairly straightforward and map easily to most UI environments, but things get really interesting when it comes to native layout.</p>
<pre name="4a49" id="4a49">var styles = StyleSheet.create({<br>  container: {<br><strong>display: 'flex'</strong><br>  }<br>})</pre>
<p name="7900" id="7900">Despite being outside of a browser environment, <em>React Native ships with its own native implementation of </em><a href="https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Using_CSS_flexible_boxes" rel="nofollow noopener" target="_blank"><em>flexbox</em></a><em>.</em></p>
<p name="6d80" id="6d80">Initially released as a JavaScript package called <a href="https://www.npmjs.com/package/css-layout" rel="nofollow noopener" target="_blank"><em>css-layout</em></a><em>,</em> reimplementing flexbox entirely in JavaScript (backed by an appropriately comprehensive test suite), it’s now been migrated to C for better portability.</p>
<p name="55d8" id="55d8">Given the scope and importance of the project, it’s been given a more significant brand of its own in the form of <a href="https://facebook.github.io/yoga" rel="nofollow noopener" target="_blank">Yoga</a>.</p>
<figure name="2161" id="2161"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*mv_hHmbOgU7SOd5t2J2Q2g.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*mv_hHmbOgU7SOd5t2J2Q2g.png">
</div>
</div>
<figcaption>Yoga (<a href="https://facebook.github.io/yoga" rel="noopener nofollow" target="_blank">Source</a>)</figcaption></figure><p name="8d6f" id="8d6f">Even though Yoga is all about porting CSS concepts to non-browser environments, the potentially unmanageable scope has been reigned in by focusing on only a subset of CSS features.</p>
<blockquote name="97ea" id="97ea">“Yoga’s focus is on creating an expressive layout library, not implementing all of CSS”</blockquote>
<p name="858a" id="858a">These sorts of tradeoffs may seem limiting, but when you look at the history of CSS architecture, it’s clear that <em>working with CSS at scale is all about picking a reasonable subset of the language.</em></p>
<p name="6a32" id="6a32">In Yoga’s case, they eschew the cascade in favour of scoped styles, and focus their layout engine entirely on flexbox. While this rules out a lot of functionality, it also unlocks an amazing opportunity for cross-platform components with embedded styling, and we’ve already seen several notable open source projects trying to capitalise on this fact.</p>
<p name="65c0" id="65c0"><a href="https://github.com/necolas/react-native-web" rel="nofollow noopener" target="_blank">React Native for Web</a> by <a href="https://twitter.com/necolas" rel="nofollow noopener" target="_blank">Nicolas Gallagher</a> aims to be a drop-in replacement for the react-native library. When using a bundler like webpack, aliasing third-party packages is fairly straightforward.</p>
<pre name="a14e" id="a14e">module: {<br>  alias: {<br>    'react-native': 'react-native-web'<br>  }<br>}</pre>
<p name="cc8b" id="cc8b">Using React Native for Web allows React Native components to work in a browser environment, including a browser port of the <a href="https://facebook.github.io/react-native/docs/stylesheet.html" rel="nofollow noopener" target="_blank">React Native StyleSheet API</a><em>.</em></p>
<p name="3ba7" id="3ba7">Similarly, <a href="https://github.com/lelandrichardson/react-primitives" rel="nofollow noopener" target="_blank">react-primitives</a> by <a href="https://twitter.com/intelligibabble" rel="nofollow noopener" target="_blank">Leland Richardson</a> is all about providing a cross-platform set of primitive components that abstract away the implementation details of the target platform, creating a workable baseline for cross-platform components.</p>
<p name="1e65" id="1e65">Even <a href="https://github.com/Microsoft" rel="nofollow noopener" target="_blank">Microsoft</a> is getting in on the act with the introduction of <a href="https://microsoft.github.io/reactxp" rel="nofollow noopener" target="_blank">ReactXP</a>, a library designed to streamline efforts to share code across both web and native, which also includes its own <a href="https://microsoft.github.io/reactxp/docs/styles.html" rel="nofollow noopener" target="_blank">platform-agnostic style implementation</a>.</p>
<p name="8a1c" id="8a1c">—</p>
<p name="d34e" id="d34e">Even if you don’t write software for native applications, it’s important to note that having a truly cross-platform component abstraction allows us to target an effectively limitless set of environments, sometimes in ways that you might never have predicted.</p>
<p name="3a5b" id="3a5b">The most surprising example of this I’ve seen is <a href="http://airbnb.io/react-sketchapp" rel="noopener nofollow" target="_blank">react-sketchapp</a> by <a href="https://twitter.com/jongold" rel="noopener nofollow" target="_blank">Jon Gold</a> at <a href="https://github.com/airbnb" rel="noopener nofollow" target="_blank">Airbnb</a>.</p>
<figure name="0255" id="0255"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*qfskIhHAWpYwfR5Lz0_cIA.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*qfskIhHAWpYwfR5Lz0_cIA.png">
</div>
</div>
<figcaption>react-sketchapp (<a href="https://github.com/airbnb/react-sketchapp" rel="nofollow noopener" target="_blank">Source</a>)</figcaption></figure><p name="977d" id="977d">For many of us, so much of our time is spent trying to standardise our design language, limiting the amount of duplication in our systems as much as possible. Unfortunately, as much as we’d like to only have a single source of truth, it seemed like the best we could hope for was to reduce it to <em>two</em>—a living style guide for developers, and<em> a static style guide for designers.</em> While certainly much better than what we’ve been historically used to, this still leaves us manually syncing from design tools—like <a href="https://www.sketchapp.com" rel="nofollow noopener" target="_blank">Sketch</a>—to code and back. This is where react-sketchapp comes in.</p>
<p name="72fd" id="72fd">Thanks to Sketch’s <a href="http://developer.sketchapp.com/reference/api" rel="nofollow noopener" target="_blank">JavaScript API</a>, and the ability for React to connect to different renderers, react-sketchapp lets us take our cross-platform React components and render them into our Sketch documents.</p>
<figure name="b1cb" id="b1cb"><div>
<div></div>
<div>
<img src="https://cdn-images-1.medium.com/freeze/max/1000/1*v2L1DB8OS38GScyBRFD8hQ.png?q=20"><canvas></canvas><img><img src="https://cdn-images-1.medium.com/max/1000/1*v2L1DB8OS38GScyBRFD8hQ.png">
</div>
</div>
<figcaption>Profile Cards example powered by react-sketchapp (<a href="http://airbnb.io/react-sketchapp/docs/examples.html" rel="nofollow noopener" target="_blank">Source</a>)</figcaption></figure><p name="d60a" id="d60a">Needless to say, this has the potential to massively shake up how designers and developers collaborate. Now, when we refer to the same components while iterating on our designs, we can potentially be referring to the same <em>implementation</em> too, regardless of whether we’re working with tools for designers or developers.</p>
<p name="d23a" id="d23a">With <a href="https://www.sketchapp.com/learn/documentation/symbols" rel="nofollow noopener" target="_blank">symbols in Sketch</a> and <a href="https://facebook.github.io/react/docs/components-and-props.html" rel="nofollow noopener" target="_blank">components in React</a>, our industry is essentially starting to converge on the same abstraction, opening the opportunity for us to work closer together by sharing the same tools.</p>
</div></div></section><section name="a0d6"><div><hr></div>
<div><div>
<p name="c723" id="c723">It’s no coincidence that so many of these new experiments are coming from within the React community, and those communities surrounding it.</p>
<p name="f2ef" id="f2ef">In a component architecture, co-locating as many of a component’s concerns in a single place becomes a high priority. This, of course, includes its locally scoped styles, but even extends to more complicated areas like data fetching thanks to libraries like <a href="https://facebook.github.io/relay" rel="nofollow noopener" target="_blank">Relay</a> and <a href="http://dev.apollodata.com" rel="nofollow noopener" target="_blank">Apollo</a>. The result is something that unlocks an enormous amount of potential, of which we’ve only just scratched the surface.</p>
<p name="7d6d" id="7d6d">While this has a huge impact on the way we style our applications, it has an equally large effect on everything else in our architecture—but for good reason.</p>
<p name="37a0" id="37a0">By unifying our model around components written in a single language, we have the potential to better separate our concerns—not by technology, but by <em>functionality.</em> Scoping everything around the unit of a component, scaling large yet maintainable systems from them, optimised in ways that weren’t possible before, sharing our work with each other more easily and composing large applications out of small open-source building blocks. Most importantly, we can do all of this without breaking progressive enhancement, without giving up on the principles that many of us see as a non-negotiable part of taking the web platform seriously.</p>
<p name="6210" id="6210">Most of all, I’m excited about the potential of components written in a single language to form the basis of <em>a new, unified styling language</em>—one that unites the front-end community in ways we’ve never seen before.</p>
</div></div></section><section name="8d2b"><div><hr></div>
<div><div>
<p name="0b70" id="0b70">At SEEK, we’re working to take advantage of this by building our own living style guide around this component model, where semantics, interactivity and visual styling are all united under a single abstraction. This forms a common design language, shared between developers and designers alike.</p>
<p name="947b" id="947b">As much as possible, building a page should be as simple as combining an opinionated set of components that ensure our work stays on brand, while allowing us to upgrade our design language long after we’ve shipped to production.</p>
<pre name="7c31" id="7c31">import <strong>{<br>  PageBlock,<br>  Card,<br>  Text<br>}</strong> from 'seek-style-guide/react'</pre>
<pre name="da32" id="da32">const App = () =&gt; (<strong><br>  &lt;PageBlock&gt;<br>    &lt;Card&gt;<br>      &lt;Text heading&gt;</strong>Hello World!<strong>&lt;/Text&gt;<br>    &lt;/Card&gt;<br>  &lt;/PageBlock&gt;<br></strong>)</pre>
<p name="ee6a" id="ee6a">Even though our style guide is currently built with React, webpack and CSS Modules, the architecture exactly mirrors what you’d find in any system built with CSS-in-JS. The technology choices may differ, but the mindset is the same.</p>
<p name="e81e" id="e81e">However, these technology choices will likely need to shift in unexpected ways in the future, which is why keeping an eye on this space is so critical to the ongoing development of our component ecosystem. We may not be using CSS-in-JS today, but it’s quite possible that a compelling reason to switch could arise sooner than we think.</p>
</div></div></section><section name="f86b"><div><hr></div>
<div><div>
<p name="f7d5" id="f7d5">CSS-in-JS has come a surprisingly long way in a short amount of time, but it’s important to note that, in the grand scheme of things, it’s only just getting started.</p>
<p name="8579" id="8579">There’s still so much room for improvement, and the innovations are showing no signs of stopping. Libraries are still popping up to address the outstanding issues and to improve the developer experience—performance enhancements, extracting static CSS at build time, targeting CSS variables and lowering the barrier to entry for all front-end developers.</p>
<p name="3ef5" id="3ef5">This is where the CSS community comes in. Despite all of these massive alterations to our workflow, <strong><em>none of this changes the fact that you still need to know CSS.</em></strong></p>
<p name="de67" id="de67">We may express it with different syntax, we may architect our apps in different ways, but the fundamental building blocks of CSS aren’t going away. Equally, our industry’s move towards component architectures is inevitable, and the desire to reimagine the front-end through this lens is only getting stronger. There is a very real need for us to work together, to ensure our solutions are widely applicable to developers of all backgrounds, whether design-focused, engineering-focused, or both.</p>
<p name="7b3d" id="7b3d">While we may sometimes seem at odds, the CSS and JS communities both share a passion for improving the front-end, for taking the web platform seriously, and improving our processes for the next generation of web sites. There’s so much potential here, and while we’ve covered an incredible amount of ground so far, there’s still so much work left to be done.</p>
</div></div></section><section name="d4a5"><div><hr></div>
<div><div>
<p name="9528" id="9528">At this point, you still might not be convinced, and that’s totally okay. It’s completely reasonable to find CSS-in-JS to be ill-fitting for your work <em>right now,</em> but my hope is that it’s for the <em>right reasons,</em> rather than superficial objections to mere <em>syntax.</em></p>
<p name="745f" id="745f">Regardless, it seems quite likely that this approach to authoring styles is only going to grow more popular over the coming years, and it’s worth keeping an eye on it while this approach continues to evolve at such a rapid pace. I sincerely hope that you’re able to join us in helping make the next generation of CSS tooling as effective as possible for all front-end developers, whether through code contributions or <em>simply being an active part of the conversation. </em>If not, at the <em>very least,</em> I hope I’ve been able to give you a better understanding of why people are so passionate about this space, and—maybe—why it’s not such a ridiculous idea after all.</p>
</div></div></section><section name="2395"><div><hr></div>
<div><div><p name="daa3" id="daa3"><em>This article was written in parallel with a talk of the same name — presented at CSSconf EU 2017 in Berlin, Germany — which is now </em><a href="https://www.youtube.com/watch?v=X_uTCnaRe94" rel="nofollow noopener" target="_blank"><em>available on YouTube</em></a><em>.</em></p></div></div></section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium https://medium.com<br>
<br>
February 20, 2019 at 01:41AM
